client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +790,116 @@ where
 	}
 }
 
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactment was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
+}
+
+impl<Block> EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	fn new() -> Self {
+		EnactmentState { recent_best_block: None, recent_finalized_block: None }
+	}
+
+	pub fn update_and_check_if_new_enactment_is_valid<PoolApi: graph::ChainApi<Block = Block>>(
+		&mut self,
+		api: &PoolApi,
+		event: &ChainEvent<Block>,
+	) -> (bool, Option<TreeRoute<Block>>) {
+		let (new_hash, finalized) = match event {
+			ChainEvent::NewBestBlock { hash, .. } => (*hash, false),
+			ChainEvent::Finalized { hash, .. } => (*hash, true),
+		};
+
+		let best_block = match self.recent_best_block {
+			Some(recent_best_hash) => recent_best_hash,
+			None => {
+				self.recent_best_block = Some(new_hash);
+				if finalized {
+					self.recent_finalized_block = Some(new_hash);
+				}
+
+				return (true, None)
+			},
+		};
+
+		// compute actual tree route from best_block to notified block, and use it instead of
+		// tree_route provided with event
+		let tree_route =
+			api.tree_route(best_block, new_hash).unwrap().expect("tree_route exists. qed.");
+
+		log::trace!(target: "txpool", "resolve hash:{new_hash:?} finalized:{finalized:?} tree_route:{tree_route:?}, best_block:{best_block:?}, finalized_block:{:?}", self.recent_finalized_block);
+
+		// block was already finalized
+		if let Some(finalized_block) = self.recent_finalized_block {
+			if finalized_block == new_hash {
+				log::trace!(target:"txpool", "handle_enactment: block already finalized: exit 3b");
+				return (false, Some(tree_route))
+			}
+
+			// check if recently finalized block is on retracted path...
+			if tree_route.retracted().iter().any(|x| x.hash == finalized_block) {
```

Comment:
Maybe I will add error-message there and leave implementation as proposed?

Reply:
After reading this again, this case can indeed happen as a result of out-of-order `NBB`. Using the blockchain state of the comment above (also from the PR description): we process `F(C1)` and then processing `NBB(C2)` would lead to this state and the handling is correct, i.e. it tries to retract a finalized block therefore this event is stale and there's nothing to do.